pkg/server/image_pull.go

```diff
@@ -17,14 +17,315 @@ limitations under the License.
+	}
+	digest := desc.Digest.String()
+
+	// TODO pass a metadata struct to pullimage
+	// store the image metadata
+	var meta *metadata.ImageMetadata
+	meta, err = c.imageMetadataStore.Get(digest)
+	if err == nil {
+		if meta == nil {
+			meta = &metadata.ImageMetadata{
+				ID:          digest,
+				RepoTags:    []string{image},       // TODO fill this array properly
+				RepoDigests: []string{digest},      // TODO fill this array properly
+				Size:        uint64(size),          // TODO  compressed or uncompressed size? using compressed
+				Auth:        r.GetAuth(),           // TODO security
+				CreatedAt:   time.Now().UnixNano(), // TODO or should this be the image created at time?
+			}
+			if err = c.imageMetadataStore.Create(*meta); err != nil {
+				return &runtime.PullImageResponse{ImageRef: digest},
+					fmt.Errorf("pulled image `%q` but failed to store metadata for digest: %s err: %v", image, digest, err)
+			}
+		}
+	}
+
+	// Return the image digest
+	return &runtime.PullImageResponse{ImageRef: digest}, err
+}
+
+// imageReferenceResolver attempts to resolve the image reference into a name
+// and descriptor via the containerd library call..
+//
+// The argument `ref` should be a scheme-less URI representing the remote.
+// Structurally, it has a host and path. The "host" can be used to directly
+// reference a specific host or be matched against a specific handler.
+//
+// The returned name should be used to identify the referenced entity.
+// Dependending on the remote namespace, this may be immutable or mutable.
+// While the name may differ from ref, it should itself be a valid ref.
+// Note: as of now revoleved name appears to be the sha256:digest
+//
+// If the resolution fails, an error will be returned.
+// TODO just starting this function...
+func (c *criContainerdService) imageReferenceResolver(ref string) (resolvedImageName string, manifest ocispec.Manifest, compressedSize uint64, err error) {
+	ctx := gocontext.Background()
+	var desc ocispec.Descriptor
+	resolvedImageName, desc, _, err = docker.NewResolver().Resolve(ctx, ref)
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize, fmt.Errorf("failed to resolve ref %q: %v", ref, err)
+	}
+	image, _ := c.imageStore.Get(ctx, desc.Digest.String())
+	p, err := content.ReadBlob(ctx, c.provider, image.Target.Digest)
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize, fmt.Errorf("failed to readblob: %v", err)
+	}
+	err = json.Unmarshal(p, &manifest)
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize, fmt.Errorf("failed to unmarshal manifest %v", err)
+	}
+	size, err := image.Size(ctx, c.provider)
+	if err != nil {
+		return resolvedImageName, manifest, compressedSize, fmt.Errorf("failed to sum size %v", err)
+	}
+	compressedSize = uint64(size)
+
+	return resolvedImageName, manifest, compressedSize, nil
+}
+
+func (c *criContainerdService) pullImage(ref string) (ocispec.Descriptor, int64, error) {
```

Comment:
We can make this function much simpler.
1) Resolve the image.
2) Pull images with `Dispath`.
3) Put the information into image store.
4) Unpack the image.

We don't need those goroutines and progress stuff.